diff --git a/preload/harness/src/ld_preload_fuzz.c b/preload/harness/src/ld_preload_fuzz.c
index 4225728..e1f4ed4 100644
--- a/preload/harness/src/ld_preload_fuzz.c
+++ b/preload/harness/src/ld_preload_fuzz.c
@@ -709,8 +709,7 @@ void* shmat(int shmid, const void* shmaddr, int shmflg) {
 
 void _exit(int status) {
   if (nyx_firefox_is_parent && status != 1337) {
-    fprintf(stderr, "[*] _exit: Attempted to escape the matrix\n");
-    *((volatile int*)0x1) = __LINE__;
+    hprintf("[*] _exit: Attempted to escape the matrix\n");
   }
 
   void (*__exit)(int status) = dlsym(RTLD_NEXT, "_exit");
